### ISSUE TITLE
Access to all errors

### ISSUE THREAD

**Issue by user471** [flycheck/flycheck#829]
Could you add variable ```flycheck-all-current-errors```. It's similar to ```flycheck-current-errors``` but with list of all errors from all files.

**Comment by lunaryorn** [comment]
@user471 What would you need that for?  What advantage would the variable provide over simply iterating over all buffers and collecting `flycheck-current-errors`?

**Comment by user471** [comment]
>What would you need that for?
When you compile the project, errors could be anywhere, not only in the current buffer.

>What advantage would the variable provide over simply iterating over all buffers and collecting flycheck-current-errors?
There could be no buffer for file with error.

**Comment by lunaryorn** [comment]
> There could be no buffer for file with error.

I'm sure that there's a misunderstanding here:  Flycheck does not check files, it checks _buffers_.  If there's no buffer for a file Flycheck won't even know about it.  In other words, there _must_ be a buffer for Flycheck to check a file.

Flycheck's purpose is to show errors for the current buffer, not for the whole project.

I recommend that you use `M-x compile` or Projectile's project compilation feature to get all errors.

**Comment by user471** [comment]
>I recommend that you use M-x compile or Projectile's project compilation feature to get all errors.

Flycheck already has everything I need. Most of the time the error in current buffer. But sometime it can be in other file, so the only missing feature is list of all errors.
I have added ```  (setq flycheck-all-current-errors errors)``` into ```flycheck-finish-current-syntax-check``` and everything works fine, so I thought maybe it could be useful for everyone.

**Comment by lunaryorn** [comment]
I'm not sure how your code looks like exactly, but if you added that expression verbatim to `flycheck-finish-current-syntax-check` then `flycheck-all-current-errors` won't contain all errors, but only those of the very last syntax check in any buffer.

You'd have to add—and consequently remove as well—errors selectively in order to maintain a global list of all errors—which is somewhat cumbersome.  And it will not give you _all_ errors in _all_ files of your project but only those of the files that you actually visit in Emacs, i.e. that have buffers.  In other words it'd be equivalent to

```
(seq-mapcat (lambda (buffer)
              (seq-copy (buffer-local-value 'flycheck-current-errors buffer)))
            (buffer-list))
```

And the list of errors that obtain this way is mostly outdated, because Flycheck only checks buffers that are actually visible in some window.

Flycheck is really not made for what you're trying to use it for, and adding a proper global list of errors would be more intricate than just adding and removing errors from a global variable.

It's beyond Flycheck's scope, so I'm closing this issue as `wontfix`.

**Comment by user471** [comment]
> And it will not give you all errors in all files of your project but only those of the files that you actually visit in Emacs

It isn't true. I just open only one file, and ```flycheck-all-current-errors``` contains all errors from other files. If defined checker return all errors, why would they disappear? Here is my code.

```
(defun flycheck-finish-current-syntax-check (errors)
  (setq flycheck-all-current-errors errors)
  (let* ((syntax-check flycheck-current-syntax-check)
         (checker (flycheck-syntax-check-checker syntax-check))
         (errors
          (flycheck-relevant-errors
                  (flycheck-fill-and-expand-error-file-names
                   (flycheck-filter-errors
                    (flycheck-assert-error-list-p errors) checker)))))
```
Before ```flycheck-relevant-errors``` is used ```errors``` variable contains all error which was returned by checker.

**Comment by lunaryorn** [comment]
@user471 That is by accident.  If a syntax checker happens to output errors for other files as well (e.g. includes in C++ for instance) Flycheck parses these errors as well, and only filters them later on.

But Flycheck does not expect syntax checkers to check other files beyond the current one, it does not tell syntax checkers to do so, and consequently the vast majority of syntax checkers included in Flycheck only check the current buffer and nothing more.  And _no_ syntax checker explicitly checks all files in a project.

In other words, if you visit a Python file you'll only ever get the errors of the current file, regardless of whether there are errors in other Python files, because Flycheck runs pylint or flake8 only on the current file, and neither tools automatically looks at other files.

**Comment by user471** [comment]
@lunaryorn It looks like Flycheck is mostly used with dynamically type languages and for them it make sense to check one file, but for statically typed languages it's not the case. You cannot just check one file, if you use a compiler you need to compile all or most of them. Maybe Flycheck could rethink something about it.
By adding one line to Flycheck I could use Flycheck with such approach.

**Comment by lunaryorn** [comment]
@user471 Flycheck works with many statically typed languages, including Haskell, Rust, C, etc.

We will stick to a buffer-based model for simple syntax checking. It works, it's easy to understand and maintain, and generally does the job well for many languages.

For more complicated use cases this model is not sufficient, for sure, but there's no simple generic model either.  In that case we will rather rely on language-specific backends.  

For instance, OCaml support uses Tuareg for checking, Flycheck only provides the display frontend.

**Comment by cpitclaudel** [comment]
@user471: Sebastian has made plenty of good points (I personally often use Flycheck with statically typed languages, and I'm not sure how static typing connects with approaches to modularity, nor with using or not using a compiler). Moreover, just because a language requires compiling multiple files to check a single one, it doesn't imply that the compiler needs to expose (or even compute) detailed error information for all of them (that would be redundant and wasteful, especially if multiple files depend on the same one).

You may be interested in [this blog post](http://endlessparentheses.com/flycheck-a-directory-and-report-the-results.html) if you want to check all files in a directory.

**Comment by user471** [comment]
> I'm not sure how static typing connects with approaches to modularity, nor with using or not using a compiler). Moreover, just because a language requires compiling multiple files to check a single one, it doesn't imply that the compiler needs to expose (or even compute) detailed error information for all of them 

Yes, you can just ignore other errors and work only with current buffer, this is how Flycheck works now. It's  just not practical. When you use dynamically typed language and you don't see the errors in you buffer, what do you do? You run your program or your tests and watch if there is something wrong. But with statically typed language, if you don't have errors in current buffer, you still need to compile you program before run. So, you use ```M-x compile``` then you go to other buffer with error, then you run Flycheck again. Obviously it isn't an optimal approach. That is why I added ```flycheck-all-current-errors``` and now I don't need to use ```M-x compile``` at all.

**Comment by lunaryorn** [comment]
@user471  Please do not jump to conclusions too fast, and please do not assume that your own expectations and experience reflects that of other users.   I'd like to ask you to change your style of writing.

Absolute claims like "It's just not practical" or "_Obviously_ it isn't an optimal approach"—apart from being generally wrong—block an open discussion and do not bring us forward.  Please maintain an open perspective and try to see the needs of other users as well.

----

I am happy to hear that Flycheck can replace `M-x compile` for you, but Flycheck does not aim to replace compilation nor does it guarantee to give the same results.  We could do that if we would always scan all files of a project, but we do not do that, not even for statically typed, compiled languages, *for a very important reason*:

**Flycheck's purpose and selling point is _instant_ feedback.**

As such Flycheck needs to be *fast* before anything else—which necessarily conflicts with scanning all files of a project, the more the larger the project gets.

Consequently we actively try to look at **as few files as possible** to scan the current buffer for errors.   Even in Haskell we only look at the modules the current buffer depends on, and we only apply hlint to the current buffer—which is absolutely sufficient to get error information for the current buffer.

In short, Flycheck favours **speed** and **instant feedback** over *precision* and *completeness*.  Having no Flycheck errors in the current buffer gives a great chance—but no guarantee—that the current buffer files, but it definitely doesn't tell you whether the whole project builds.

Even with Flycheck, you _have to build_ your project frequently.

And by the way, even statically typed compiled languages usually have tests which you need to run :smiley:

**Comment by user471** [comment]
>apart from being generally wrong—block an open discussion and do not bring us forward

How can I say it differently? If there are the errors in other files it better to know about this errors then not to know. How can It be wrong? Yes, if a language has tool which can check only one file then everything is fine, but almost every not mainstream statically typed language has only compiler and it returns all errors in all files. So, for such languages there isn't any benefits from one buffer approach. 
I don't insist on anything. Maybe this languages is not suitable for Flycheck, but yet again, I added one line of code and everything now work fine.

>Flycheck's purpose and selling point is instant feedback.
I use Flycheck only on save(again because I have only compiler) and I am pretty happy without instant feedback. For me most important features are highlight of error location and popup with error message.

**Comment by lunaryorn** [comment]
> How can I say it differently?

Please make less absolute claims, try to use the word "I" and phrase statements from your perspective, try to provide evidence or concrete examples and use cases, and try to see our perspective. 

> If a language has tool which can check only one file then everything is fine, but almost every not mainstream statically typed language has only compiler and it returns all errors in all files. 

I'm sure there are plenty of linters for non-mainstream languages as well, depending on your definition of "mainstream".  This statement does not really help us.  

Please, if you have a  specific issue with a particular language name it. We can try to find a solution that improves Flycheck's support for that language.  But we will not add features to Flycheck based on purely theoretical arguments, the less if these features conflict with Flycheck's core features.

Instant feedback is one of these core features.  You may not need it, that is great, but nonetheless many users want it and like it.  Please understand and accept that your point of view in this aspect is not at all a majority view point, let alone an authoritative opinion.

Collecting "all" errors a syntax checker outputs conflicts with our goal of instant feedback.  We cannot guarantee what errors a syntax checker outputs, let alone that it outputs any errors from other files at all, unless we try to guarantee that a syntax checker looks at other files as well, which definitely impedes instant feedback, for the longer compilation times alone.  

But without that guarantee it's just an "accidental" feature, because it'd be entirely unspecified what files a syntax checker looks at, and the absence of any errors would still not say anything about the state of the whole project.

I'm sorry, but as things stand I see no way to reasonably implement this feature in Flycheck.   If you have a concrete issue with a specific language, please open a new issue and we will try to help you solve the issue, but I see no point for us to continue this discussion.

**Comment by mgudemann** [comment]
@user471 I wrote a checker for a non-mainstream, i.e., limited user base, statically-typed language, see https://github.com/flycheck/flycheck-mercury This language has an `error-check only` mode where a module is only analyzed but not compiled and uses the information inside interface files to get the information about imported functions from other modules.

In general, having the information about the signature of modules should be enough for statically typed languages. Of course, this may be out of date, but then I simply regenerate the files. This should work analogously in Ocaml with its `.mli` files and Haskell with the `.hi` files. Maybe you should tell us which language you are using, so that we can try to find a good solution.

As @lunaryorn said, the main goal of Flycheck is being fast, not necessarily complete; it should help to quickly find errors in the visited file.

**Comment by user471** [comment]
@mgudemann For example, you have two files: ```foo.m``` is opened in Emacs and  ```bar.m``` is not opened in Emacs. You have changed something in ```foo.m``` and it has broken something in file ```bar.m```. Flycheck shows no erros. How do you find out about file ```bar.m```?

**Comment by mgudemann** [comment]
@user471 I regenerate the files, just as I wrote before. This is also in the documentation for `flycheck-mercury` https://github.com/flycheck/flycheck-mercury#usage

The point is to limit the amount of compilation / analysis going on in the background, also there a no files created by Flycheck in this case. In my opinion, this outweighs the disadvantage that the interface files may be out of date.

**Comment by user471** [comment]
>I regenerate the files, just as I wrote before.

This is my point. Instead of additional operation I just want to see all current errors.

>The point is to limit the amount of compilation / analysis going on in the background, also there a no files created by Flycheck in this case. In my opinion, this outweighs the disadvantage that the interface files may be out of date.

The compiler could support incremental compilation so there would not be any disadvantage. Flycheck could rely on compiler and only show errors from the checker without any analysis. 
Maybe Flymake could work like this?

**Comment by lunaryorn** [comment]
> The compiler could support incremental compilation so there would not be any disadvantage. Flycheck could rely on compiler (like thin client) and only show errors from the checker without any analysis.

I am sorry but that is not a direction Flycheck will head to in the near or distant future.

Maybe, if #774 is implemented, we might one day have a mode that runs `M-x compile` in background and annotates the current buffer with the resulting errors, but that is only an vague idea and not a goal of Flycheck, let alone a priority.

If you have any concrete patches that you would like to add to Flycheck to build your own extension on top of Flycheck we will however consider them independently from this discussion.  Please feel free to open a pull request in this case.

Thank you.

> Maybe Flymake could work like this?

I do not think so.  Flymake is even less flexible than Flycheck in this aspect (in all aspects generally).

**Comment by shajra** [comment]
I think this issue got caught up in the idea of doing a global compilation, which I understand is not in the direction of what Flycheck is trying to offer.  I have other means of doing that sort of thing.  But earlier there was this idea by @lunaryorn:

> ```
> (seq-mapcat (lambda (buffer)
>     (seq-copy (buffer-local-value 'flycheck-current-errors buffer)))
>     (buffer-list))
> ```
> And the list of errors that obtain this way is mostly outdated, because Flycheck only checks buffers that are actually visible in some window.

I think I'm actually okay this this.  Sometimes I have two buffers open, and they have interactions where corrections in one buffer correct problems in another.  I don't often use split windows, so for now, I have to remember the names of the buffers that have problems.  Even if the list of errors were outdated, I'm okay navigating to each buffer to see if I can get the list of errors clear again.  I understand this is no proof that my build is in good shape -- just a heuristic for the few buffers I have open; if I want more, I need to run a proper build across the whole project.

So I'm guessing the answer is "okay, if that's what you want, you can do that."  My only problem is that my Elisp isn't so great yet, and looking at the Flycheck source code I haven't yet figured out the best way to get this `seq-mapcat` call to update the "*Flycheck errors*" buffer instead of the current code that updates it for just the current focused buffer.

**Comment by fmdkdd** [comment]
@shajra I think the `*Flycheck errors*` buffer is unique, and tied to the buffer you are currently viewing.  If you are collecting errors from all buffers in a `flycheck-all-current-errors` variable, you should be able to substitute that variable for `flycheck-current-errors` wherever you want the global list.

Also, take a look at #1191, where we give meaning to errors from other files collected from a single compilation.  It is a bit different than merging errors from multiple compilations, but there is some overlap if, e.g., your checker *already* returned errors for multiple files but Flycheck previously chose to ignore all errors for files other than the one you are visiting in the active buffer.

**Comment by CodeSwimBikeRunner** [comment]
I honestly read through this whole thing, and have no idea where this all ended up.

Flycheck to me is a linter, a linter for specifically checking the current file. 

But I could also see linting all files even if this is outside of the flycheck world.
Is there some command I can create that uses flycheck to iterate over all buffers. Any examples would be great, as I am no emacs expert. The seq-mapcat solution works? How would I bind that to a keybinding with flycheck?

**Comment by cpitclaudel** [comment]
> Is there some command I can create that uses flycheck to iterate over all buffers. 

Sure, that's what the seq-mapcat thing does (it gathers a list of all flycheck errors in all buffers.

> How would I bind that to a keybinding with flycheck?

Check `define-key`


I thik you might be interested in lsp-mode, which in some cases checks more than one file
